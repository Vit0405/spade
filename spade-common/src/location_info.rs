use std::ops::Range;

use codespan::Span;

use crate::error_reporting::AsLabel;

pub trait HasCodespan {
    fn codespan(&self) -> Span;
}
impl<T> HasCodespan for Loc<T> {
    fn codespan(&self) -> Span {
        self.span
    }
}
impl HasCodespan for Span {
    fn codespan(&self) -> Span {
        self.clone()
    }
}
impl HasCodespan for std::ops::Range<usize> {
    fn codespan(&self) -> Span {
        lspan(self.clone())
    }
}

pub trait WithLocation: Sized {
    fn at(self, file_id: usize, span: &impl HasCodespan) -> Loc<Self>
    where
        Self: Sized,
    {
        Loc::new(self, span.codespan(), file_id)
    }

    /// Creates a new Loc from another Loc
    fn at_loc<T: Sized>(self, loc: &Loc<T>) -> Loc<Self> {
        Loc::new(self, loc.span, loc.file_id)
    }

    fn between(
        self,
        file_id: usize,
        start: &impl HasCodespan,
        end: &impl HasCodespan,
    ) -> Loc<Self> {
        Loc::new(self, start.codespan().merge(end.codespan()), file_id)
    }

    fn between_locs<T, Y>(self, start: &Loc<T>, end: &Loc<Y>) -> Loc<Self> {
        assert!(start.file_id == end.file_id);
        Loc::new(self, start.codespan().merge(end.codespan()), end.file_id())
    }

    fn nowhere(self) -> Loc<Self>
    where
        Self: Sized,
    {
        self.at(0, &Span::new(0, 0))
    }
}

impl WithLocation for () {}
impl WithLocation for u128 {}
impl WithLocation for u64 {}
impl WithLocation for usize {}
impl WithLocation for bool {}
impl<T> WithLocation for Vec<T> {}

pub fn lspan(s: logos::Span) -> Span {
    Span::new(s.start as u32, s.end as u32)
}

#[cfg(test)]
pub fn dummy() -> Span {
    Span::new(0, 0)
}

#[derive(Clone, Copy)]
pub struct Loc<T> {
    pub inner: T,
    pub span: Span,
    pub file_id: usize,
}

impl<T> Loc<T> {
    pub fn new(inner: T, span: Span, file_id: usize) -> Self {
        Self {
            inner,
            span,
            file_id,
        }
    }
    pub fn nowhere(inner: T) -> Self {
        Self::new(inner, Span::new(0, 0), 0)
    }

    pub fn strip(self) -> T {
        self.inner
    }

    pub fn strip_ref(&self) -> &T {
        &self.inner
    }

    pub fn separate(self) -> (Self, Span) {
        let span = self.span;
        (self, span)
    }

    pub fn separate_loc(self) -> (Self, Loc<()>) {
        let loc = self.loc();
        (self, loc)
    }

    pub fn split(self) -> (T, Span) {
        (self.inner, self.span)
    }
    pub fn split_ref(&self) -> (&T, Span) {
        (&self.inner, self.span)
    }
    pub fn split_loc(self) -> (T, Loc<()>) {
        let loc = self.loc();
        (self.inner, loc)
    }
    pub fn split_loc_ref(&self) -> (&T, Loc<()>) {
        let loc = self.loc();
        (&self.inner, loc)
    }

    pub fn map<Y>(self, mut op: impl FnMut(T) -> Y) -> Loc<Y> {
        Loc {
            inner: op(self.inner),
            span: self.span,
            file_id: self.file_id,
        }
    }
    pub fn map_ref<Y>(&self, mut op: impl FnMut(&T) -> Y) -> Loc<Y> {
        Loc {
            inner: op(&self.inner),
            span: self.span,
            file_id: self.file_id,
        }
    }
    pub fn try_map_ref<Y, E, F>(&self, mut op: F) -> Result<Loc<Y>, E>
    where
        F: FnMut(&T) -> Result<Y, E>,
    {
        Ok(Loc {
            inner: op(&self.inner)?,
            span: self.span,
            file_id: self.file_id,
        })
    }

    pub fn loc(&self) -> Loc<()> {
        Loc {
            inner: (),
            span: self.span,
            file_id: self.file_id,
        }
    }
}

impl<T> Into<Range<usize>> for Loc<T> {
    fn into(self) -> Range<usize> {
        self.span.into()
    }
}

impl<T, E> Loc<Result<T, E>> {
    pub fn map_err<E2>(self, err_fn: impl Fn(E, Loc<()>) -> E2) -> Result<Loc<T>, E2> {
        match self.inner {
            Ok(inner) => Ok(Loc {
                inner,
                span: self.span,
                file_id: self.file_id,
            }),
            Err(e) => Err(err_fn(e, ().at(self.file_id, &self.span))),
        }
    }
}

impl<T> PartialEq for Loc<T>
where
    T: PartialEq,
{
    fn eq(&self, other: &Self) -> bool {
        self.inner == other.inner
    }
}

impl<T> Eq for Loc<T> where T: Eq {}

impl<T> std::fmt::Display for Loc<T>
where
    T: std::fmt::Display,
{
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.inner)
    }
}

impl<T> std::hash::Hash for Loc<T>
where
    T: std::hash::Hash,
{
    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
        self.inner.hash(state)
    }
}

impl<T> std::fmt::Debug for Loc<T>
where
    T: std::fmt::Debug,
{
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "l({:3}..{:3})[{:?}]",
            self.span.start(),
            self.span.end(),
            self.inner
        )
    }
}

impl<T> std::ops::Deref for Loc<T> {
    type Target = T;

    fn deref(&self) -> &Self::Target {
        &self.inner
    }
}
impl<T> std::ops::DerefMut for Loc<T> {
    fn deref_mut(&mut self) -> &mut Self::Target {
        &mut self.inner
    }
}

impl<T> AsLabel for Loc<T> {
    fn file_id(&self) -> usize {
        self.file_id
    }

    fn span(&self) -> std::ops::Range<usize> {
        self.span.into()
    }
}
