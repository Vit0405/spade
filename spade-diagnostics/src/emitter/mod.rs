use codespan_reporting::term::termcolor::Buffer;

use crate::diagnostic::Diagnostic;
use crate::CodeBundle;

pub use codespan_emitter::{codespan_config, CodespanEmitter};

pub mod codespan_emitter;
mod panik;

/// Something that can format and emit diagnostics.
pub trait Emitter {
    /// Emit a diagnostic, e.g. print it.
    fn emit_diagnostic(&mut self, diag: &Diagnostic, buffer: &mut Buffer, code: &CodeBundle);
}

// Glue so we can stuff our fancy new diagnostics in a thiserror. This should be removed when we
// don't have any Error-enums left.

impl std::fmt::Display for Diagnostic {
    fn fmt(&self, _: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        unimplemented!()
    }
}

impl std::error::Error for Diagnostic {}
