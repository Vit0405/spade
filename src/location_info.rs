use codespan::Span;

pub trait WithLocation {
    fn at(self, span: Span) -> Loc<Self>
    where
        Self: Sized,
    {
        Loc::new(self, span)
    }

    fn nowhere(self) -> Loc<Self>
    where
        Self: Sized,
    {
        self.at(Span::new(0, 0))
    }
}

impl WithLocation for () {}

pub fn lspan(s: logos::Span) -> Span {
    Span::new(s.start as u32, s.end as u32)
}

#[cfg(test)]
pub fn dummy() -> Span {
    Span::new(0, 0)
}

#[derive(Clone, Debug, Copy)]
pub struct Loc<T> {
    pub inner: T,
    pub span: Span,
}

impl<T> Loc<T> {
    pub fn new(inner: T, span: Span) -> Self {
        Self { inner, span }
    }
    pub fn nowhere(inner: T) -> Self {
        Self::new(inner, Span::new(0, 0))
    }

    pub fn strip(self) -> T {
        self.inner
    }

    pub fn separate(self) -> (Self, Span) {
        let span = self.span;
        (self, span)
    }

    pub fn split(self) -> (T, Span) {
        (self.inner, self.span)
    }
    pub fn split_ref(&self) -> (&T, Span) {
        (&self.inner, self.span)
    }

    pub fn map<Y>(self, mut op: impl FnMut(T) -> Y) -> Loc<Y> {
        Loc {
            inner: op(self.inner),
            span: self.span,
        }
    }
    pub fn map_ref<Y>(&self, mut op: impl FnMut(&T) -> Y) -> Loc<Y> {
        Loc {
            inner: op(&self.inner),
            span: self.span,
        }
    }

    pub fn loc(&self) -> Loc<()> {
        Loc {
            inner: (),
            span: self.span,
        }
    }
}

impl<T, E> Loc<Result<T, E>> {
    pub fn map_err<E2>(self, err_fn: impl Fn(E, Loc<()>) -> E2) -> Result<Loc<T>, E2> {
        match self.inner {
            Ok(inner) => Ok(Loc {
                inner,
                span: self.span,
            }),
            Err(e) => Err(err_fn(e, ().at(self.span))),
        }
    }
}

impl<T> PartialEq for Loc<T>
where
    T: PartialEq,
{
    fn eq(&self, other: &Self) -> bool {
        self.inner == other.inner
    }
}
