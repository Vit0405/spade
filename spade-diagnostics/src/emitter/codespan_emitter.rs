use codespan_reporting::diagnostic::{
    Diagnostic as CodespanDiagnostic, Subdiagnostic as CodespanSubdiagnostic, SuggestionPart,
};
use codespan_reporting::term::termcolor::{Color, ColorChoice, ColorSpec};
use codespan_reporting::term::{self, termcolor::Buffer};

use spade_common::location_info::AsLabel;

use crate::diagnostic::Subdiagnostic;
use crate::{CodeBundle, Diagnostic, Emitter};

pub fn color_choice(no_color: bool) -> ColorChoice {
    if no_color {
        ColorChoice::Never
    } else {
        ColorChoice::Auto
    }
}

pub fn codespan_config() -> codespan_reporting::term::Config {
    let mut primary_label_error = ColorSpec::new();
    primary_label_error
        .set_fg(Some(Color::Red))
        .set_intense(true);

    let style = codespan_reporting::term::Styles {
        primary_label_error,
        ..Default::default()
    };
    codespan_reporting::term::Config {
        styles: style,
        ..Default::default()
    }
}

pub struct CodespanEmitter;

impl Emitter for CodespanEmitter {
    fn emit_diagnostic(&mut self, diag: &Diagnostic, buffer: &mut Buffer, code: &CodeBundle) {
        let severity = diag.level.severity();
        let message = diag.message.as_str();
        let primary_label = if let Some(primary_label_message) = diag.primary_label.as_ref() {
            diag.span
                .primary_label()
                .with_message(primary_label_message.as_str())
        } else {
            diag.span.primary_label()
        };
        let mut labels = vec![primary_label];
        labels.extend(
            diag.secondary
                .iter()
                .map(|(sp, msg)| sp.secondary_label().with_message(msg.as_str())),
        );
        let mut simple_notes = vec![];
        let mut subdiagnostics = vec![];
        for subdiag in &diag.subdiagnostics {
            match subdiag {
                Subdiagnostic::Note { level, message } => {
                    simple_notes.push(format!("{}: {}", level.as_str(), message.as_str()))
                }
                Subdiagnostic::SpannedNote {
                    level,
                    message,
                    primary_label: (primary_span, primary_message),
                    secondary_labels,
                } => {
                    let mut labels = vec![primary_span
                        .primary_label()
                        .with_message(primary_message.as_str())];
                    labels.extend(
                        secondary_labels
                            .iter()
                            .map(|(sp, msg)| sp.secondary_label().with_message(msg.as_str())),
                    );
                    subdiagnostics.push(CodespanSubdiagnostic::Note {
                        severity: level.severity(),
                        message: message.as_str().to_string(),
                        labels,
                    });
                }
                Subdiagnostic::Suggestion { parts, message } => {
                    subdiagnostics.push(CodespanSubdiagnostic::Suggestion {
                        file_id: parts[0].0 .1,
                        message: message.as_str().to_string(),
                        parts: parts
                            .iter()
                            .map(|((sp, _), replacement)| SuggestionPart {
                                range: (*sp).into(),
                                replacement: replacement.to_string(),
                            })
                            .collect(),
                    })
                }
            }
        }
        let diag = CodespanDiagnostic::new(severity)
            .with_message(message)
            .with_labels(labels)
            .with_notes(simple_notes)
            .with_subdiagnostics(subdiagnostics);

        term::emit(buffer, &codespan_config(), &code.files, &diag).unwrap();
    }
}
