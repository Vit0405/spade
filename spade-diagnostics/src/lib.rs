use std::io::Write;

use codespan_reporting::files::{Files, SimpleFiles};
use codespan_reporting::term::termcolor::Buffer;

use spade_common::location_info::{AsLabel, Loc};

pub use diagnostic::Diagnostic;
pub use emitter::Emitter;

pub mod diagnostic;
pub mod emitter;

/// A bundle of all the source code included in the current compilation
#[derive(Clone)]
pub struct CodeBundle {
    pub files: SimpleFiles<String, String>,
}

impl CodeBundle {
    // Create a new code bundle adding the passed string as the 0th file
    pub fn new(string: String) -> Self {
        let mut files = SimpleFiles::new();
        files.add("<str>".to_string(), string);
        Self { files }
    }

    pub fn from_files(files: &[(String, String)]) -> Self {
        let mut result = Self {
            files: SimpleFiles::new(),
        };
        for (name, content) in files {
            result.add_file(name.clone(), content.clone());
        }
        result
    }

    pub fn add_file(&mut self, filename: String, content: String) -> usize {
        self.files.add(filename, content)
    }

    pub fn dump_files(&self) -> Vec<(String, String)> {
        let mut all_files = vec![];
        loop {
            match self.files.get(all_files.len()) {
                Ok(file) => all_files.push((file.name().clone(), file.source().clone())),
                Err(codespan_reporting::files::Error::FileMissing) => break,
                Err(e) => {
                    panic!("{e}")
                }
            };
        }
        all_files
    }

    pub fn source_loc<T>(&self, loc: &Loc<T>) -> String {
        let location = self
            .files
            .location(loc.file_id(), loc.span.start().to_usize())
            .expect("Loc was not in code bundle");
        format!(
            "{}:{},{}",
            self.files.get(loc.file_id()).unwrap().name(),
            location.line_number,
            location.column_number
        )
    }
}

pub trait CompilationError {
    fn report(&self, buffer: &mut Buffer, code: &CodeBundle, diag_handler: &mut DiagHandler);
}

impl CompilationError for std::io::Error {
    fn report(&self, buffer: &mut Buffer, _code: &CodeBundle, _diag_handler: &mut DiagHandler) {
        if let Err(e) = buffer.write_all(self.to_string().as_bytes()) {
            eprintln!(
                "io error when writing io error to error buffer\noriginal error: {}\nnew error: {}",
                self, e
            );
        }
    }
}

pub struct DiagHandler {
    emitter: Box<dyn Emitter + Send>,
    // Here we can add more shared state for diagnostics. For example, rustc can
    // stash diagnostics that can be retrieved in later stages, indexed by (Span, StashKey).
}

impl DiagHandler {
    pub fn new(emitter: Box<dyn Emitter + Send>) -> Self {
        Self { emitter }
    }

    pub fn emit(&mut self, diagnostic: &Diagnostic, buffer: &mut Buffer, code: &CodeBundle) {
        self.emitter.emit_diagnostic(diagnostic, buffer, code);
    }
}

#[cfg(test)]
mod tests {
    use codespan_reporting::term::termcolor::Buffer;

    use crate::emitter::CodespanEmitter;
    use crate::{CodeBundle, Diagnostic, Emitter};

    #[test]
    fn bug_diagnostics_works() {
        let code = CodeBundle::new("hello goodbye".to_string());
        let sp = ((6..13).into(), 0);
        let mut buffer = Buffer::no_color();
        let mut emitter = CodespanEmitter;
        let diagnostic = Diagnostic::bug(sp, "oof");
        emitter.emit_diagnostic(&diagnostic, &mut buffer, &code);
        insta::assert_display_snapshot!(String::from_utf8(buffer.into_inner()).unwrap());
    }
}
