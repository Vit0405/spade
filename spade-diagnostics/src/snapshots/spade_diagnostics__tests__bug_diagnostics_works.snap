---
source: spade-diagnostics/src/lib.rs
expression: "String::from_utf8(buffer.into_inner()).unwrap()"
---
bug: oof
  ┌─ <str>:1:7
  │
1 │ hello goodbye
  │       ^^^^^^^
  │
  = note: This is an internal bug in the compiler.
          We would appreciate if you opened an issue in the repository:
          https://gitlab.com/spade-lang/spade/-/issues/new?issuable_template=Internal%20bug


